### PR TITLE
Measure force inference

### PR DESCRIPTION
**Pull request by sophietheis**
Measure force inference in the tyssue according to the method describe in Brodland et al.

### CODE REVIEW

File: tyssue/core/sheet.py [hunk 1]
```diff
@@ -273,6 +275,75 @@ def extract_bounding_box(
...
+        n_edges = len(edges_index)
+
+        srce, trgt = self.edge_df.loc[
+            edges_index, ['srce', 'trgt']].iteritems()
+        srce = srce[1].to_numpy()
+        trgt = trgt[1].to_numpy()
+
+        # Fill gamma matrix to measure tension
+        pos = (self.edge_df.loc[edges_index, dcoords].to_numpy() / np.repeat(np.linalg.norm(
+            self.edge_df.loc[edges_index, dcoords].to_numpy(), axis=1), ndim).reshape((n_edges, ndim)))
```
Comment (glyg) [inline tyssue/core/sheet.py:307]: This is computed in self.edge_df[["ux", "uy", "uz"]]] but this is done [here](https://github.com/DamCB/tyssue/blob/master/tyssue/dynamics/factory.py#L112) - which is not really good. Can you use `ucoords` here and move the computation in `factory.py` to `base_geometry`?

File: tyssue/core/sheet.py [hunk 2]
```diff
@@ -273,6 +275,75 @@ def extract_bounding_box(
...
+         où C1 = {1....1}
+
+        """
+        if coords is None:
+            coords = self.coords
+        ndim = self.dim
+        dcoords = ["d" + c for c in coords]
+
+        self.get_extra_indices()
+        edges_index = self.east_edges
```
Comment (glyg) [inline tyssue/core/sheet.py:295]: I think you can use `sgle_edges` here to do the computation also for the free edges at the border?

File: tyssue/core/sheet.py [hunk 3]
```diff
@@ -273,6 +275,75 @@ def extract_bounding_box(
...
+        pos = np.concatenate((pos, -pos)).flatten()
+        row = np.concatenate((
+            np.array([np.arange(i * ndim, i * ndim + ndim)
+                      for i in srce]).flatten(),
+            np.array([np.arange(i * ndim, i * ndim + ndim)
+                      for i in trgt]).flatten())
+        )
+
+        col = np.concatenate(
+            (np.repeat(np.arange(n_edges), ndim), np.repeat(np.arange(n_edges), ndim)))
```
Comment (glyg) [inline tyssue/core/sheet.py:318]: If I understand correctly you solve for both east and west edges, but it should be possible to only solve for west_edges (plus free_edges maybe) and set the east_edges tension (or its opposite) for the west_edes?

File: tyssue/core/sheet.py [hunk 4]
```diff
@@ -273,6 +275,75 @@ def extract_bounding_box(
...
+        r1, r2 = linalg.qr(p)
+        y = np.dot(r1.T, q)
+
+        x = linalg.solve(r2, y)
+
+        tension = x[0:n_edges][:, 0]
+
+        self.edge_df['tension'] = np.nan
+        self.edge_df.loc[edges_index, 'tension'] = tension
+        self.edge_df.loc[edges_index_opposite, 'tension'] = tension
```
Comment (glyg) [inline tyssue/core/sheet.py:345]: maybe it's better to return the column and let the user assign it (or, alternatively set the column name as a parameter in the function call)

File: tyssue/core/sheet.py [hunk 5]
```diff
@@ -273,6 +275,75 @@ def extract_bounding_box(
...
+        Equation is homogenous and to avoid tension_vectors = 0,
+        Construction and solve the constrained least-squares equation system
+        [[g_gamma_matrix.T g_gamma_matrix, C^T_1],
+         [C_1, 0]]
+         où C1 = {1....1}
+
+        """
+        if coords is None:
+            coords = self.coords
+        ndim = self.dim
```
Comment (glyg) [inline tyssue/core/sheet.py:291]: maybe `ndim` should be `len(coords)` here (if we do strange projections for ex.).

File: tyssue/core/sheet.py (hunk 3 above)
Comment (sophietheis) [reply tyssue/core/sheet.py:318]: I resolve for east edges only, and after set the east_edges tension for the west_edges.

File: tyssue/core/sheet.py [hunk 6]
```diff
@@ -273,6 +275,75 @@ def extract_bounding_box(
...
+        dcoords = ["d" + c for c in coords]
+
+        self.get_extra_indices()
+        edges_index = self.east_edges
+        edges_index_opposite = self.west_edges
+
+        n_edges = len(edges_index)
+
+        srce, trgt = self.edge_df.loc[
+            edges_index, ['srce', 'trgt']].iteritems()
```
[the remark this answers is not in the document]
Comment (sophietheis) [reply tyssue/core/sheet.py:301]: It is changed.

File: tyssue/core/sheet.py (hunk 1 above)
Comment (sophietheis) [reply tyssue/core/sheet.py:307]: As I understand, I add a method `update_ucoords` in base geometry?

Comment (glyg) [reply tyssue/core/sheet.py:307]: yes please :)

File: tyssue/core/sheet.py (hunk 3 above)
Comment (glyg) [reply tyssue/core/sheet.py:318]: but then why is the matrix this size? Maybe I miss something here

Comment (sophietheis) [reply tyssue/core/sheet.py:318]: The matrix size is (Ne/2, Nv*ndim)

File: tyssue/core/sheet.py (hunk 1 above)
Comment (sophietheis) [reply tyssue/core/sheet.py:307]: I spend so much time to this line, and finally I just have to use `self.edge_df[["ux", "uy", "uz"]]]`...
Pourquoi faire simple quand on peut faire compliqué... --'

Comment (glyg) [reply tyssue/core/sheet.py:307]: :smile:

File: tyssue/core/sheet.py (hunk 3 above)
Comment (glyg) [reply tyssue/core/sheet.py:318]: ok, so there is indeed something I did not get :)

one thing that might make things easier (and drop the list comprehensions) is to arrange the data a bit differently. Now you do (if I understand correctly) 

[x_s0, y_s0, z_s0, x_s1, y_s1, z_s1, ... x_si, y_si, z_si, ...]. where si is the index of the source vertex for edge i. Now if you use [x_s0, x_s1 ... x_si ..., y_s0, y_s1, ...], you can write  your rows as 
```py
row = np.concatenate(
    [srce+i*Ne for i in range(dim)] + [trgt+i*Ne for i in range(dim)]
)
```

and the cols are "just" `np.tile(np.arange(n_edges), dim*2))` or something approaching

Comment (sophietheis) [reply tyssue/core/sheet.py:318]: I don't have the same result with this method. Matrix is not fill with the same values in both method. 
Also, I think it's important to keep x,y,z coordinates together to solve the matrix after.

File: tyssue/core/sheet.py (hunk 2 above)
Comment (sophietheis) [reply tyssue/core/sheet.py:295]: I will use `sgle_edges` only if the user want to take into account free border edges.

File: tyssue/geometry/base_geometry.py [hunk 7]
```diff
@@ -37,6 +37,12 @@ def update_dcoords(sheet):
         else:
             update_periodic_dcoords(sheet)
 
+    @staticmethod
```
Comment (glyg) [inline tyssue/geometry/base_geometry.py:40]: can you also add a call to the method in update_all(), and remove the call in factory.py?

File: tyssue/core/sheet.py [hunk 8]
```diff
@@ -288,40 +288,37 @@ def get_force_inference(self, coords=None):
...
+        ndim = len(coords)
+        ucoords = ["u" + c for c in coords]
 
         self.get_extra_indices()
-        edges_index = self.east_edges
-        edges_index_opposite = self.west_edges
+        edges_index = self.east_edges.to_numpy()
+        edges_index_opposite = self.west_edges.to_numpy()
+        if free_border_edges:
+            edges_lonely = np.setxor1d(self.sgle_edges, edges_index)
```
Comment (glyg) [inline tyssue/core/sheet.py:298]: I think those are self.free_edges - they are already computed

File: tyssue/geometry/bulk_geometry.py [hunk 9]
```diff
@@ -38,6 +39,10 @@ def update_all(cls, eptm):
     def update_dcoords(eptm):
         SheetGeometry.update_dcoords(eptm)
 
+    @staticmethod
+    def update_ucoords(eptm):
+        SheetGeometry.update_ucoords(eptm)
+
```
Comment (glyg) [inline tyssue/geometry/bulk_geometry.py:45]: I don't know why it's there, inheritance should be enough (both for `update_dcoords` and `update_ucoords`), could you just delete those lines?

**Comment by glyg** [comment]
just one very last (and easy) thing: could you add an entry to changelog.md to say you added a force inference algo? Thanks :D

**Pull request merged**